File: lua/emu/cpu.lua
local cpu
cpu = {
	pc = 0,
	actual_memory = {},
	total_cycles = 0,

	-- Memory proxy allows 0-based access to the 1-based lua table
	memory = setmetatable({}, {
		__index = function(_, k)
			return cpu.actual_memory[k + 1]
		end,
		__newindex = function(_, k, v)
			cpu.actual_memory[k + 1] = v
		end
	}),

	subleq = function()
		if (cpu.pc < 0) then return 0 end

		local addr_A = cpu.memory[cpu.pc]
		local addr_B = cpu.memory[cpu.pc + 1]
		local addr_C = cpu.memory[cpu.pc + 2]

		-- mechanism for printing a character to the screen
		if (addr_B == -1) then
			local val_a = cpu.memory[addr_A]
			io.write(string.char(val_a))

			-- this is to help check for the end of a string via a negative value
			if (-val_a <= 0) then
				cpu.pc = addr_C
			else
				cpu.pc = cpu.pc + 3
			end
		else
			local val_a = cpu.memory[addr_A]
			local val_b = cpu.memory[addr_B]
			local result = val_b - val_a
			cpu.memory[addr_B] = result

			if (result <= 0) then
				cpu.pc = addr_C
			else
				cpu.pc = cpu.pc + 3
			end
		end

		return 1
	end,

	step = function()
		cpu.total_cycles = cpu.total_cycles + cpu.subleq()
	end,

	run = function(cycles)
		while (cpu.total_cycles < cycles and cpu.pc >= 0) do
			cpu.step()
		end
	end,

	load_program = function(input)
		cpu.actual_memory = input
		cpu.pc = 0
		cpu.total_cycles = 0
	end,

	reset = function()
		cpu.pc = 0
		cpu.total_cycles = 0
	end,
}
return cpu

File: go/emu/cpu.go
package emu

import (
	"fmt"
)

type CPU struct {
	PC          int
	Memory      []int
	TotalCycles int
}

func NewCPU() *CPU {
	return &CPU{
		PC:          0,
		Memory:      nil,
		TotalCycles: 0,
	}
}

func (cpu *CPU) Subleq() int {
	if cpu.PC < 0 || cpu.PC+2 >= len(cpu.Memory) {
		return 0
	}

	addrA := cpu.Memory[cpu.PC]
	addrB := cpu.Memory[cpu.PC+1]
	addrC := cpu.Memory[cpu.PC+2]

	if addrB == -1 {
		if addrA >= 0 && addrA < len(cpu.Memory) {
			valA := cpu.Memory[addrA]
			fmt.Printf("%c", valA)

			if -valA <= 0 {
				cpu.PC = addrC
			} else {
				cpu.PC += 3
			}
		} else {
			return 0
		}
	} else {
		if addrA >= 0 && addrA < len(cpu.Memory) && addrB >= 0 && addrB < len(cpu.Memory) {
			valA := cpu.Memory[addrA]
			valB := cpu.Memory[addrB]
			result := valB - valA
			cpu.Memory[addrB] = result

			if result <= 0 {
				cpu.PC = addrC
			} else {
				cpu.PC += 3
			}
		} else {
			return 0
		}
	}

	return 1
}

func (cpu *CPU) Step() {
	cpu.TotalCycles += cpu.Subleq()
}

func (cpu *CPU) Run(cycles int) {
	for cpu.TotalCycles < cycles && cpu.PC >= 0 {
		cpu.Step()
	}
}

func (cpu *CPU) LoadProgram(input []int) {
	cpu.Memory = make([]int, len(input))
	copy(cpu.Memory, input)
	cpu.PC = 0
	cpu.TotalCycles = 0
}

func (cpu *CPU) Reset() {
	cpu.PC = 0
	cpu.TotalCycles = 0
}
